scheduler/util.go

```diff
@@ -169,121 +172,158 @@ func shuffleNodes(plan *structs.Plan, index uint64, nodes []*structs.Node) {
 	}
 }
 
-// tasksUpdated does a diff between task groups to see if the
-// tasks, their drivers, environment variables or config have updated. The
-// inputs are the task group name to diff and two jobs to diff.
-// taskUpdated and functions called within assume that the given
-// taskGroup has already been checked to not be nil
-func tasksUpdated(jobA, jobB *structs.Job, taskGroup string) bool {
+// comparison records the _first_ detected difference between two groups during
+// a comparison in tasksUpdated
+//
+// This is useful to provide context when debugging the result of tasksUpdated.
+type comparison struct {
+	modified bool
+	label    string
+	before   any
+	after    any
+}
+
+func difference(label string, before, after any) comparison {
+	// push string formatting into String(), so that we never call it in the
+	// hot path unless someone adds a log line to debug with this result
+	return comparison{
+		modified: true,
+		label:    label,
+		before:   before,
+		after:    after,
+	}
+}
+
+func (c comparison) String() string {
+	return fmt.Sprintf("%s changed; before: %#v, after: %#v", c.label, c.before, c.after)
+}
+
+// same indicates no destructive difference between two task groups
+var same = comparison{modified: false}
+
+// tasksUpdated creates a comparison between task groups to see if the tasks, their
+// drivers, environment variables or config have been modified.
+func tasksUpdated(jobA, jobB *structs.Job, taskGroup string) comparison {
```

Comment:
> Since we return early on first change detected I think it would make sense to defer comparison for less used fields

I thought about this, and introduced this [Benchmark](https://github.com/hashicorp/nomad/blob/main/scheduler/util_test.go#L20) in case someone wants to improve things. For this PR I really only care about correctness. 

Unfortunately no generic hashing or equals methodology will work here - there's a handful of custom comparisons we need to do (i.e. ignoring or normalizing fields) even outside of `Canonicalize`, which you can see in the helper methods called by `tasksUpdated`.

Reply:
Even if we could generate a hash, we'd probably want to cache that hash value as part of the struct for it to benefit performance. Otherwise we'd have to rehash on each pass through the scheduler, which would mean chasing all the same set of pointers we need to do to call the helper methods anyways. That's why ex. the node class hash is precalculated.

Reply:
> I thought about this, and introduced this [Benchmark](https://github.com/hashicorp/nomad/blob/main/scheduler/util_test.go#L20) in case someone wants to improve things. For this PR I really only care about correctness.

Makes sense to focus on correctness. I think the problem with benchmarking something like this is that it's not an empirical performance change, it's more like a heuristic based on what we _think_ is most often modified. 

Just for a quick check I modified the benchmark so it updates a task `env` and moved the task loop to the top and got some improvement:

```
goos: darwin
goarch: arm64
pkg: github.com/hashicorp/nomad/scheduler
BenchmarkTasksUpdated-10          125265              9232 ns/op
BenchmarkTasksUpdated-10          138152              8697 ns/op
BenchmarkTasksUpdated-10          128814              8714 ns/op
BenchmarkTasksUpdated-10          136887              8711 ns/op
BenchmarkTasksUpdated-10          138432              8690 ns/op
BenchmarkTasksUpdated-10          137956              8736 ns/op
BenchmarkTasksUpdated-10          136660              8736 ns/op
BenchmarkTasksUpdated-10          138224              8728 ns/op
BenchmarkTasksUpdated-10          137574              8729 ns/op
BenchmarkTasksUpdated-10          133634              8733 ns/op
PASS
ok      github.com/hashicorp/nomad/scheduler    13.029s
```

```
goos: darwin
goarch: arm64
pkg: github.com/hashicorp/nomad/scheduler
BenchmarkTasksUpdated-10          238004              4909 ns/op
BenchmarkTasksUpdated-10          237406              4527 ns/op
BenchmarkTasksUpdated-10          327616              3654 ns/op
BenchmarkTasksUpdated-10          324906              3646 ns/op
BenchmarkTasksUpdated-10          325503              3808 ns/op
BenchmarkTasksUpdated-10          325220              3822 ns/op
BenchmarkTasksUpdated-10          320337              3730 ns/op
BenchmarkTasksUpdated-10          279225              3694 ns/op
BenchmarkTasksUpdated-10          321788              3704 ns/op
BenchmarkTasksUpdated-10          317722              3680 ns/op
PASS
ok      github.com/hashicorp/nomad/scheduler    12.379s
```